src/input/type_record.cc

```diff
@@ -228,45 +229,21 @@ bool Record::check_key_default_value(const Default &dflt, const TypeBase &type,
 
 
 
-bool Record::finish()
+bool Record::finish(bool is_generic)
 {
 	if (data_->finished) return true;
 
 	ASSERT(data_->closed_, "Finished Record '%s' must be closed!", this->type_name().c_str());
 
-	// remove duplicate Abstracts in vector of parent pointers
-	if ( data_->parent_ptr_.size()>1 ) {
-	    /* Possible simplification using std library:
- 	     * sort( vec.begin(), vec.end() );
-         * vec.erase( unique( vec.begin(), vec.end() ), vec.end() );
-         *
-         * ... needs sort with particular comparator.
-	     */
-		for (auto it = data_->parent_ptr_.begin(); it != data_->parent_ptr_.end(); ++it) {
-			TypeHash hash = (*it)->content_hash();
-			for (auto in_it = it+1; in_it != data_->parent_ptr_.end(); ++in_it) {
-				if ( (*in_it)->content_hash() == hash ) { // same parent - remove
-					data_->parent_ptr_.erase( in_it );
-					--in_it;
-				}
-			}
-		}
-	}
-
-	// finish inheritance if parent is non-null
-    if ( data_->parent_ptr_.size() ) {
-    	ASSERT( data_->keys.size() > 0 && data_->keys[0].key_ == "TYPE",
-    				"Derived record '%s' must have defined TYPE key!\n", this->type_name().c_str() );
-    	boost::shared_ptr<TypeBase> type_copy = boost::make_shared<Selection>( data_->parent_ptr_[0]->get_type_selection() );
-    	data_->keys[0].type_ = type_copy;
-    	data_->keys[0].default_ = Default( type_name() );
-    }
-
     data_->finished = true;
     for (vector<Key>::iterator it=data_->keys.begin(); it!=data_->keys.end(); it++)
     {
-        if (it->key_ != "TYPE") {
-            data_->finished = data_->finished && const_cast<TypeBase *>( it->type_.get() )->finish();
+    	if (it->key_ != "TYPE") {
```

Comment:
OK, setting  type  of the key TYPE to NULL in derive_from is error prone and needs checks like this one. What about to set it to an empty selection. Then we still can distinguish completed TYPE keys if it is necessary but make it litle bit more safe. This change however needs careful check of all places dealing with the type of TYPE.